tests: improves tapscript unit tests

EthanHeilman opened the pull request:
This commit creates new test utilities for future Taproot script tests within script_tests.json. The key features of this commit are the addition of three new tags: `#SCRIPT#`, `#CONTROLBLOCK#`, and `#TAPROOTOUTPUT#`. These tags streamline the test creation process by eliminating the need to manually generate these components outside the test suite.

* `#SCRIPT#`: Parses Tapscript and outputs a byte string of opcodes.
* `#CONTROLBLOCK#`: Automatically generates the control block for a given Taproot output.
* `#TAPROOTOUTPUT#`: Generates the final Taproot scriptPubKey.

This code was originally part of the OP_CAT PR https://github.com/bitcoin/bitcoin/pull/29247 but was pulled out into a separate PR to reduce the rebase treadmill for the OP_CAT PR.

Additionally this PR adds a check to ensure that if the witness data can not be parsed as hex the test fails. Prior to this PR, the test code would fail silently and set the values it couldn't parse as empty stack elements. This fix was suggested by @instagibbs.

## Rationale 

While writing JSON script tests (script_tests.json) for https://github.com/bitcoin/bitcoin/pull/29247 we ran into the following problem. The JSON script tests are simple and easy to write for pre-Tapscript scripts, but adding or changing a Tapscript test requires substantial work per test. Consider the following pre-tapscript test:

```
["'aa' 'bb'", "CAT 0x4c 0x02 0xaabb EQUAL", "P2SH,STRICTENC", "DISABLED_OPCODE", "CAT disabled"]
````

whereas a Tapscript test for the same script (annotated with comments for better readability) would look like:

```
[
    [
        "aa",
        "bb",
        "7e4c02aabb87", // output script
        "c0d6889cb081036e0faefa3a35157ad71086b123b2b144b649798b494c300a961d", // control block
        0.00000001
    ],
    "",
    "0x51 0x20 0x15048ed3a65748549c27b671936987093cf73a4c9cb18522a74fb9553060ca99", // Tapscript output
    "P2SH,WITNESS,TAPROOT",
    "OK",
    "TAPSCRIPT CATs aa and bb together and checks if EQUAL to aabb"
]
```

Computing the Tapscript output, such as `0x51 0x20 0x15048ed3a65748549c27b671936987093cf73a4c9cb18522a74fb9553060ca99`, requires writing custom code and running it for each test. The same is true for the Tapscript control block, such as `c0d6889cb081036e0faefa3a35157ad71086b123b2b144b649798b494c300a961d`. If a test is changed or updated new outputs and control blocks must be computed. The complexity of doing this is likely the reason that no one has added any Tapscript tests to JSON script tests until this PR.

In this PR we address this issue by adding the following improvements to JSON script tests:

Adding simple macros ("#SCRIPT# and #CONTROLBLOCK#) that allow the script test parser to automatically generate and inject a valid Tapscript output and control block to be computed automatically from the JSON script.
Allowing Tapscript scripts to use the human readable strings like pre-script scripts by marking the location of the script in the witness stack using #SCRIPT#. This transforms the unreadable script 7e4c02aabb87 into #SCRIPT# CAT 0x4c 0x02 0xaabb EQUAL.
This results in the following JSON script test which is far easier to write and easier to read.

```
[
    [
        "aa",
        "bb",
        "#SCRIPT# CAT",
        "#CONTROLBLOCK#",
        0.00000001
    ],
    "",
    "0x51 0x20 #TAPROOTOUTPUT#",
    "P2SH,WITNESS,TAPROOT,OP_CAT",
    "OK",
    "TAPSCRIPT Test of OP_CAT flag by calling CAT on two elements. TAPSCRIPT_OP_CAT flag is set so CAT is executed."
],
```

instagibbs commented:
is this in draft for a reason?

EthanHeilman commented:
> is this in draft for a reason?

@instagibbs Yes, my todos to make this not a draft are:

[ ] - Better PR description
[ ] - A few extra tapscript tests (it was most cat tapscript tests before)
[ ] - Doublechecking my code to make sure the rebase didn't break anything

I got most of this done over the weekend, I just haven't pushed it yet. Will likely push it tonight, Wednesday

0xBEEFCAF3 commented on src/test/data/script_tests.json:
```diff
@@ -2610,6 +2610,63 @@
+    "0x51 0x20 #TAPROOTOUTPUT#",
+    "P2SH,WITNESS,TAPROOT",
+    "OK",
+    "TAPSCRIPT Tests testing tapscript with many different op codes including ALTSTACK interactions. "
```
Super nit: empty space on this line. Not sure if the linter even picks up on this

EthanHeilman replied:
Fixed! Thanks

0xBEEFCAF3 commented on src/test/script_tests.cpp:
```diff
@@ -940,7 +958,14 @@ BOOST_AUTO_TEST_CASE(script_json_test)
+        CScript scriptPubKey;
+        // If requested, auto-generate the taproot output
+        if (strcmp(scriptPubKeyString.c_str(), "0x51 0x20 #TAPROOTOUTPUT#")== 0) {
+            BOOST_CHECK_MESSAGE(taprootBuilder.IsComplete(), "Failed to autogenerate Tapscript Script PubKey");
```
Nit: I would use "taproot output key"  here instead of "Tapscript Script Pubkey" per [BIP341](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki)

EthanHeilman replied:
Fixed!

arminsabouri reviewed (approved):
tACK 1bd2af1de6551d67c1c363b4cd04d4973e3650bf
Tested by rebasing the new [bip347 tapscript unit tests](https://github.com/bitcoin/bitcoin/pull/29247) on top of these changes.

dergoegge commented:
Concept ACK

I would suggest that you [squash](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits) your commits (as you are fixing up the first commit in the second one).

Just as a note, there's also `test/functional/feature_taproot.py` which produces [taproot unit tests](https://raw.githubusercontent.com/bitcoin-core/qa-assets/refs/heads/main/unit_test_data/script_assets_test.json). I'm not sure if those were meant to be extended in the case of Tapscript updates but that seems a bit daunting anyway compared to what you suggest here.

EthanHeilman commented:
Squashed.

> Just as a note, there's also `test/functional/feature_taproot.py` which produces [taproot unit tests](https://raw.githubusercontent.com/bitcoin-core/qa-assets/refs/heads/main/unit_test_data/script_assets_test.json). I'm not sure if those were meant to be extended in the case of Tapscript updates but that seems a bit daunting anyway compared to what you suggest here.

Yes, the idea here is to have the nice script unittest tests work for Tapscript. These tests are much more lightweight and are fast and easy to write.

dergoegge commented:
I think this is a good addition to the unit test framework but I'd like to see some more tapscript specific test cases added in this PR as well (e.g. OP_SUCCESSx opcodes, schnorr sigs, no check multi sig, ...).

EthanHeilman commented:
@dergoegge 
I agree and I'd like to add those tests in a second PR if this PR is merged?

I don't have a lot of time this month and next to work on them and I worry about putting in more time for a PR that may not be merged.

EthanHeilman commented:
@instagibbs  Any changes you want me to make?

instagibbs commented:
Let me take a deeper look

EthanHeilman committed:
test: Ensures test fails if witness is not hex

This commit ensures that we do not fail silently when the test script encounters a witness string in the JSON test data that can not be parsed as hex.

EthanHeilman committed:
test: improves tapscript unit tests

This commit creates new test utilities for future Taproot script
tests within script_tests.json. The key features of this commit are the
addition of three new tags: `#SCRIPT#`, `#CONTROLBLOCK#`, and
`#TAPROOTOUTPUT#`. These tags streamline the test creation process by
eliminating the need to manually generate these components outside the
test suite.

* `#SCRIPT#`: Parses Tapscript and outputs a byte string of opcodes.
* `#CONTROLBLOCK#`: Automatically generates the control block for a given
Taproot output.
* `#TAPROOTOUTPUT#`: Generates the final Taproot scriptPubKey.

Update src/test/script_tests.cpp

Co-authored-by: Jan B <608446+janb84@users.noreply.github.com>

darosior commented:
> The complexity of doing this is likely the reason that no one has added any Tapscript tests to JSON script tests until this PR.

I think the reason no one has added more tests is because the Tapscript logic is already extensively covered through the JSON script unit tests (in additional to the extensive functional test). See this unit test (which works with [this JSON file](https://raw.githubusercontent.com/bitcoin-core/qa-assets/refs/heads/main/unit_test_data/script_assets_test.json) from the qa-assets repo):
https://github.com/bitcoin/bitcoin/blob/804675930598ccd3102eab642d2b5aa3caa982d4/src/test/script_tests.cpp#L1605

EthanHeilman commented:
@sipa 

This PR doesn't contain any improvements to the python functional tests because they are excellent and did not require modifications to write clear tapscript tests.  This PR is not intended to replace more complex tests. It is only intended to make simple unit tests easy to write and read.

In defense of simple unittests for tapscript:

1. **JSON script_test is the lowest effort way to add a test** Adding a JSON script_test for everything but tapscript is trivially easy. Given human nature you'll get more tests if adding a test takes 5 minutes with an 1 hour learning curve than if it takes 1 hour with 1 week learning curve. I want to make it as easy as possible for someone who encounters something unexpected in a tapscript to write a test for it.

2. **script_test is expansive** As a consequence of JSON script_test being easy to add they cover many cases. We caught bugs early in development because there was an existing simple test that immediately failed. There was one case in particular with counter-intuitive (to me) script interpreter behavior that only script test caught. (behavior around false OP_IF branch data pushes)

3. **Fail faster**  Failing in unittests rather failing in functional tests results in faster cycles.

4. **Easy to debug** It is easier to debug C++ code from inside a C++ unit tests rather than python functional tests. I often use script_tests to explore script interpreter behavior. I replace script_test.json with the script I want to test, set a breakpoint and run.

I've found this workflow of debugging with script_tests valuable to me that I plan to extend it and provide a way to specify a script test at runtime via an argument rather than compile time via the json file. I know this is possible with some of the fuzzing harnesses but it isn't pleasant. The script_test format is easier to work with for everything but tapscript and with this PR tapscript becomes easy to work with.

EthanHeilman commented:
@darosior 

The static test cases are valuable, but they are multistep process to create and are hard to read. Imagine you write some code and get a  test failure. 

Would you rather the test failure be:

```JSON 
{"tx": "f705d6e8019870958e85d1d8f94aa6d74746ba974db0f5ccae49a49b32dcada4e19de4eb5ecb00000000925977cc01f9875c000000000016001431d2b00cd4687ceb34008d9894de84062def14aa05406346", "prevouts": ["b4eae1010000000022512039f7e9232896f8100485e38afa652044f855e734a13b840a3f220cbd5d911ad5"], "index": 0, "success": {"scriptSig": "", "witness": ["25e45bd4d4b8fcd5933861355a2d376aad8daf1af1588e5fb6dfcea22d0d809acda6fadca11e97f5b5c85af99df27cb24fa69b08fa6c790234cdc671d3af5a7302"]}, "failure": {"scriptSig": "", "witness": [""]}, "flags": "P2SH,DERSIG,CHECKLOCKTIMEVERIFY,CHECKSEQUENCEVERIFY,WITNESS,NULLDUMMY,TAPROOT", "final": true, "comment": "siglen/empty_keypath"},
```

or that and this more readable unit test

```JSON
[
    [
        "abcdef",
        "#SCRIPT# 1 IF SHA256 ENDIF SIZE SWAP DROP 32 EQUAL",
        "#CONTROLBLOCK#",
        0.00000001
    ],
    "",
    "0x51 0x20 #TAPROOTOUTPUT#",
    "P2SH,WITNESS,TAPROOT",
    "OK",
    "TAPSCRIPT Test IF conditional when true"
],
```

instagibbs commented:
> Would you rather the test failure be:

If you "walk" backwards to the functional test via the `comment` field it's ok, but I agree I would rather the tests cases be declared in the functional test rather than the dumped qa-assets version which lacks the context.

With regards to debugging, I found both this and https://github.com/bitcoin/bitcoin/pull/32114 similar in pain to review/fix, but I admit I've gotten used to gdb-attaching to bitcoind for functional tests regardless.

instagibbs reviewed (approved):
LGTM dfea1fae92a1fd8ba0bbf5d52a37baffa28f9f8f

I think having "fail-fast", developer-friendly tests is a fine motivation. And nothing stops me/you from porting novel things to the functional framework for additional coverage.

The additional sanity check added in first commit also helps protect against test regressions in the future.

Hand-checked the scripts, seems correct.

janb84 commented on src/test/script_tests.cpp:
```diff
@@ -916,16 +916,39 @@ BOOST_AUTO_TEST_CASE(script_json_test)
+                    const auto witness_value{TryParseHex<unsigned char>(element)};
+                    if (!witness_value.has_value()) {
+                        BOOST_ERROR("Bad witness in test: " << strTest << " witness is not hex: " << element);
+                        assert(witness_value.has_value());
```
Found this construction bid odd, asserting something that is known to fail. If the assertion is removed the unit test fails more gracefully, from 234 failures reported to 2 failures reported. Is this approach deliberate?


```suggestion
```

janb84 reviewed (commented):
ACK [b544d1c](https://github.com/bitcoin/bitcoin/pull/31640/commits/b544d1c1aeb701cc7895fa38bf3f0745806c64fd)

- Code review ✅
- Tested ✅

instagibbs commented:
could you squash commits?

EthanHeilman commented:
> could you squash commits?

@instagibbs 
Squashed

janb84 reviewed (commented):
Re ACK [62b6071](https://github.com/bitcoin/bitcoin/commit/62b6071fd6ac2aad2fc42f135b3df662ab0eee7a)

EthanHeilman commented:
This is good to merge from my perspective, let me know if anyone wants any additional changes?

EthanHeilman commented:
Any reason this hasn't been merged despite 2 approvals?

glozow commented:
I only see 1 ack, ping @instagibbs @0xBEEFCAF3 for re-review

instagibbs commented:
reACK 62b6071fd6ac2aad2fc42f135b3df662ab0eee7a

removed the assert in favor of the single BOOST_ERROR check

janb84 reviewed (commented):
Re ACK [e3d7533](https://github.com/bitcoin/bitcoin/pull/31640/commits/e3d7533ac954516d87a09b230c2898f5c9b1e213)

Changes since last ACK:
- code cleanup

instagibbs commented:
reACK e3d7533ac954516d87a09b230c2898f5c9b1e213
